### PR TITLE
adds basic movement, double jump, ghost

### PR DESCRIPTION
**Pull request by anranlee99**

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -1,46 +1,42 @@
...
+            jump: this.input.keyboard!.addKey('W'),
+            left: this.input.keyboard!.addKey('A'),
+            right: this.input.keyboard!.addKey('D'),
+        }
         this.keys.jump = this.input.keyboard!.addKey('W');
         this.keys.left = this.input.keyboard!.addKey('A');
         this.keys.right = this.input.keyboard!.addKey('D');
         this.keys.jump?.on('down', () => {
             this.player?.jump();
         })
```
Comment (s5bug) [inline src/main.ts]: Don't need `?` here, use `!` instead.

Instead of `.jump()` on the player, just set a variable `isJumpHeld` to `true` when `jump` is pressed and `false` when it isn't. That way the player state can be entirely contained in the player.

File: src/main.ts [hunk 2]
```diff
@@ -50,8 +46,8 @@ class TestScene extends Phaser.Scene {
         this.physics.world.setBounds(0, 0, this.w, this.h);
         this.player.setCollideWorldBounds(true);
     }
-    addGhost(){
-        this.ghost = this.add.sprite(this.w / 2, this.h / 2, 'player').setScale(0.5).setAlpha(0.25);
+    addGhost() {
+        this.ghost = this.add.sprite(this.player!.x, this.player!.y, 'player').setScale(0.5).setAlpha(0.25);
```
Comment (s5bug) [inline src/main.ts]: Shouldn't need `!` here.

File: src/main.ts [hunk 3]
```diff
@@ -65,35 +61,58 @@ class TestScene extends Phaser.Scene {
             loop: true
         })
     }
-    create() {
-        this.setUp();
-
-        //get rid of this if you don't want the ghost
-        this.addGhost();
-        
+    checkPlayerState(){
```
Comment (s5bug) [inline src/main.ts:64]: This should be a method on the player sprite, not on the scene

File: src/player.ts [hunk 4]
```diff
@@ -1,32 +1,79 @@
...
         super(scene, x, y, image)
         this.jumpCount = 2;
         this.body = new Phaser.Physics.Arcade.Body(scene.physics.world, this)
-        this.inAir = false;
+        this.state = PlayerState.IN_AIR;
+        this.movingLeft = false;
+        this.movingRight = false;
     }
 
     jump() {
```
Comment (s5bug) [inline src/player.ts]: This should be handled as part of the player sprite's state machine update; the player sprite should be checking what the inputs are during the state machine update, enter an `IN_AIR` state, and set the velocity.

File: src/player.ts [hunk 5]
```diff
@@ -1,32 +1,79 @@
...
         if (this.jumpCount) {
             this.setVelocityY(-1000);
```
Comment (s5bug) [inline src/player.ts]: Make this a constant

File: src/player.ts [hunk 6]
```diff
@@ -1,32 +1,79 @@
...
             this.jumpCount--;
+            if (this.state == PlayerState.LEFT_SLIDE) {
+                this.setVelocityX(Player.TOP_SPEED);
```
Comment (s5bug) [inline src/player.ts:41]: Possibly `SLIDING_SPEED`?

File: src/player.ts [hunk 7]
```diff
@@ -1,32 +1,79 @@
...
+            }
+
+
         }
     }
     update() {
-        if (this.body.bottom != this.scene.game.config.height) {
-            this.inAir = true;
+        //xor the moving because if youre pressing both we want 0 velocity
+        if ((this.movingLeft && !this.movingRight) || (this.movingRight && !this.movingLeft)) {
```
Comment (s5bug) [inline src/player.ts]: Can be simplified to `this.movingLeft != this.movingRight`

File: src/player.ts [hunk 8]
```diff
@@ -1,32 +1,79 @@
...
+            }
+
         } else {
-            this.inAir = false;
+            this.setVelocityY(this.body.velocity.y + Player.GRAVITY);
+            this.setVelocityX(0);
+        }
+
+        if (this.state == PlayerState.ON_GROUND || this.state == PlayerState.LEFT_SLIDE || this.state == PlayerState.RIGHT_SLIDE) {
             this.jumpCount = 2;
```
Comment (s5bug) [inline src/player.ts]: Make `MAX_JUMPS` a constant

**Review by s5bug — CHANGES_REQUESTED** [review]
Lookin sick so far!

File: src/main.ts [hunk 9]
```diff
@@ -1,12 +1,118 @@
...
+        this.physics.world.setBounds(0, 0, this.w, this.h);
+        this.player.setCollideWorldBounds(true);
+    }
+    addGhost() {
+        this.ghost = this.add.sprite(this.player!.x, this.player!.y, 'player').setScale(0.5).setAlpha(0.25);
+        this.time.addEvent({
+            delay: 1,
+            callback: () => {
+                let x = this.player?.x;
+                let y = this.player?.y;
```
Comment (s5bug) [inline src/main.ts]: Shouldn't need `?` here

File: src/main.ts [hunk 10]
```diff
@@ -1,12 +1,118 @@
...
+                this.time.delayedCall(500, () => {
+                    this.ghost!.x = x as number;
+                    this.ghost!.y = y as number;
```
Comment (s5bug) [inline src/main.ts]: Shouldn't need `as number` or `!` here

**Review by s5bug — CHANGES_REQUESTED** [review]

File: src/main.ts [hunk 11]
```diff
@@ -1,12 +1,103 @@
...
+    addGhost() {
+        this.ghost = this.add.sprite(this.player!.x, this.player!.y, 'player').setScale(0.5).setAlpha(0.25);
+        this.time.addEvent({
+            delay: 1,
+            callback: () => {
+                let x = this.player?.x;
+                let y = this.player?.y;
+                this.time.delayedCall(500, () => {
+                    this.ghost.x = x as number;
+                    this.ghost.y = y as number;
```
Comment (s5bug) [inline src/main.ts]: Shouldn't need `as number` here

**Review by s5bug — CHANGES_REQUESTED** [review]
Just a few more typescript things to fix! Other than that it looks perfect

**Comment by s5bug** [comment]
Check the CI failure: something was changed somewhere that shouldn't have been

File: vite.config.js [hunk 12]
```diff
@@ -10,8 +10,16 @@ export default defineConfig(({ command, mode }) => {
...
+            rollupOptions: {
+              input: {
+                app: './core-gameplay-1.html',
+              },
+            },
+          },
+          server: {
+            open: '/core-gameplay-1.html',
+          },
         base: basePath
```
Comment (s5bug) [inline vite.config.js]: Revert these changes

**Review by s5bug — CHANGES_REQUESTED** [review]

File: src/player.ts (hunk 6 above)
Comment (anranlee99) [reply src/player.ts:41]: discussed. this is for bouncing inthe opposite direction
